### ISSUE TITLE
incomplete recapitation with 0.400 (0.314 is fine)

### ISSUE THREAD

**Issue by grahamgower** [tskit-dev/pyslim#71]
SLiM script to generate `test.trees`:
[stdpopsim-homsap.txt](https://github.com/tskit-dev/pyslim/files/4393506/stdpopsim-homsap.txt)

```
$ slim -s 1234 stdpopsim-homsap.txt
// Initial random seed:
1234

// RunInitializeCallbacks():
initializeTreeSeq();
initializeMutationRate(0);
initializeMutationType(1, 0.5, "f", 0);
initializeGenomicElementType(1, m1, 1);
initializeGenomicElement(g1, 0, 249249);
initializeRecombinationRate(2.29712e-07, 249249);

// Starting run at generation <start>:
1
```
```
$ ls -l test.trees
-rw-r--r-- 1 grg grg 176252 Mar 27 13:38 test.trees
```
```
$ cat recap.py
import msprime
import pyslim
ts = pyslim.load("test.trees")
ts = ts.recapitate(recombination_rate=1e-8)
one_root = all(tree.num_roots == 1 for tree in ts.trees())
print(f"msprime.__version__={msprime.__version__}")
print(f"pyslim.__version__={pyslim.__version__}")
print(f"one_root={one_root}")

$ python recap.py
/home/grg/.local/lib/python3.8/site-packages/pyslim/slim_tree_sequence.py:105: UserWarning: This is an version 0.3 SLiM tree sequence. When you write this out, it will be converted to version 0.4.
  warnings.warn("This is an version {} SLiM tree sequence.".format(provenance.file_version) +
msprime.__version__='0.7.4'
pyslim.__version__='0.400'
one_root=False

```
```
$ pip install --user pyslim==0.314
...

$ python recap.py
msprime.__version__='0.7.4'
pyslim.__version__='0.314'
one_root=True
```

**Comment by grahamgower** [comment]
Simpler, but equivalent, slim script.

```
initialize() {
    defineConstant("chromosome_length", 249250);

    initializeTreeSeq();
    initializeMutationRate(0);
    initializeMutationType("m1", 0.5, "f", 0);
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, chromosome_length-1);
    initializeRecombinationRate(1e-8);
}

1 {
    sim.addSubpop(0, 1000);
}

100 {
    sim.treeSeqOutput("test.trees");
}
```

**Comment by petrelharp** [comment]
Huh, I can't replicate that, with msprime `HEAD` at least:
```
peter@skua:~/debug/pyslim/roots$ python3 recap.py 
msprime.__version__=0.7.5.dev112+ng3ee1551
pyslim.__version__=0.400
one_root=True
```

**Comment by petrelharp** [comment]
Ah-ha, but rolling back to msprime 0.7.4 does recreate the error. Strange. 
```
peter@skua:~/debug/pyslim/roots$ python3 recap.py 
msprime.__version__=0.7.4
pyslim.__version__=0.400
one_root=False
```

**Comment by petrelharp** [comment]
Ah-ha. The problem is that current development msprime switched from simulating in genetic units to real units. As a result, the `num_loci` argument to `msprime.simulate( )` is deprecated, so I removed it. For some reason, recapitating with pre-1.0 (i.e., pre-6c08fcf2dd01c04f85391323a89fe9dd1bca303c) msprime without the `num_loci` argument resulted in incomplete recapitation. This seems like a bug in msprime, but a no longer relevant one.

To be clear, these things work:
- msprime at or after [the switch to 1.0](https://github.com/tskit-dev/msprime/commit/6c08fcf2dd01c04f85391323a89fe9dd1bca303c), with no `num_loci`, ie doing `msprime.simulate(from_ts=ts, recombination_map=msprime.RecombinationMap([0.0, ts.sequence_length], [1e-8, 0.0]))`
- msprime before[the switch to 1.0](https://github.com/tskit-dev/msprime/commit/6c08fcf2dd01c04f85391323a89fe9dd1bca303c), with `num_loci`, ie doing `msprime.simulate(from_ts=ts, recombination_map=msprime.RecombinationMap([0.0, ts.sequence_length], [1e-8, 0.0], num_loci=int(ts.sequence_length)))`
but *this* does *not* work:
- msprime before [the switch to 1.0](https://github.com/tskit-dev/msprime/commit/6c08fcf2dd01c04f85391323a89fe9dd1bca303c), with *no* `num_loci`

I'm not sure what to do now, though - require development msprime? put a conditional that builds the recombination map in different ways depending on the msprime version? revert to using `num_loci` until msprime1.0 is released?

**Comment by grahamgower** [comment]
I think we should be able to use `hasattr(msprime.RecombinationMap, "discrete")` to condition on the msprime num_loci/discrete feature change, rather than conditioning on the version. At least, that was my intention in #72. But I admit, I didn't give it much testing.

**Comment by petrelharp** [comment]
Good idea.

**Comment by petrelharp** [comment]
Thanks for the report! Fixed in 1e560747fcb310afa2e37c65a985ee7b9c587f44.

**Comment by grahamgower** [comment]
Awesome! Thanks @petrelharp!
